File: q.py
#!/usr/bin/env python3

import time

start = time.time()

states = {}
final_result = 0

def update_state(name, is_bid, val, count):
    global final_result
    v = states.setdefault(name, {})
    key = val * 2
    if is_bid:
        key = key + 1
    # if count > 0:
    #     key = '+'
    # else:
    #     key = '-'
    # key = key + str(val)
    old = v.get(key, 0)
    new = old + count
    if old > 0 and new <= 0:
        del v[key]
    else:
        v[key] = new
        final_result = (final_result * 3 + new) & 0xffffffff

def run():
    class Random:
        def __init__(self):
            self.state = 0
        def next(self):
            self.state = (self.state * 11035 + 12345) & 0xffff
            return self.state
    random = Random()

    for i in range(0, 10000000):
    #for i in range(0, 10):
        v = random.next()
        s_index = v % 1000
        is_bid = (random.next() % 2) == 0
        val = random.next() % 20
        count = (random.next() % 10) - 5
        update_state(s_index, is_bid, val, count)

run()
end = time.time()

print(f"Time taken: {end - start:.2f} seconds")
print(f"Final result: {final_result}")

File: q.cpp
#include <iostream>
#include <chrono>
#include <string>
#include <unordered_map>

void run(std::uint32_t &final_result) {
    struct Random {
        std::uint32_t state = 0;

        Random() = default;

        std::uint32_t next() {
            state = (state * 11035 + 12345) & 0xffff;
            return state;
        }
    };
    struct State {
        std::unordered_map<std::string, unsigned int> v;
    };
    std::unordered_map<unsigned int, State> states;
    Random random;

    auto update_state = [&](unsigned int name, bool is_bid, unsigned int val, int count) {
        auto &v = states[name];
        std::string key = count > 0 ? "+" : "-";
        key = key + std::to_string(val);
        auto old = v.v.find(key) != v.v.end() ? v.v[key] : 0;
        auto new_val = old + count;
        if (old > 0 && new_val <= 0) {
            v.v.erase(key);
        } else {
            v.v[key] = new_val;
            final_result = (final_result * 3) + (std::uint32_t)new_val;
        }
    };
    for(auto i = 0; i < 10000000; ++i) {
        auto v = random.next();
        auto s_index = v % 1000;
        auto is_bid = (random.next() % 2) == 0;
        auto val = random.next() % 20;
        auto count = (int)(random.next() % 10) - 5;
        update_state(s_index, is_bid, val, count);
    }
}
int main()
{
    std::uint32_t final_result     = 0;
    auto start = std::chrono::high_resolution_clock::now();
    run(final_result);
    auto end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> elapsed = end - start;
    std::cout << "Elapsed time: " << elapsed.count() << " seconds" << std::endl;
    std::cout << "Final result: " << final_result << std::endl;

    return 0;
}